File: apps/kira-studio/frontend/src/views/console/planParsers/sqlite.ts
import { rollupIssues } from '../planIssues';
import type { PlanNode, QueryPlan } from '../planModel';

// P18 (v1.1) D13/D14/D15/D16: SQLite's `EXPLAIN QUERY PLAN` — F14. N rows × 4 columns
// (id, parent, notused, detail), no cost and no row estimate whatsoever, so `estimatedRowsRead`
// and `overThreshold` stay unset/false here (D14) — the settings panel says the threshold does
// not apply on this dialect rather than showing a silent zero. `parent` is 0 for nearly every row
// (nesting only appears for a subquery/compound query), so this builds a real id→children tree
// rather than assuming a flat list.
//
// sqlite.org/eqp.html: "The output format may change between SQLite releases. Applications
// should not depend on the output format of the EXPLAIN QUERY PLAN command." — taken as binding
// (D16): `detail` is matched on a small set of leading tokens for issue detection and otherwise
// shown verbatim, never re-rendered from a deeper parse.
export interface SqliteExplainRow {
  id: number;
  parent: number;
  detail: string;
}

const RELATION_RE = /^(?:SCAN|SEARCH) (\S+)/;

function nodeForRow(detail: string): PlanNode {
  const relation = RELATION_RE.exec(detail)?.[1];
  const node: PlanNode = { label: detail, relation, detail, metrics: [], issues: [], children: [] };

  // D15: full scan — SCAN (as opposed to SEARCH, which narrows via an index or rowid).
  if (detail.startsWith('SCAN ')) {
    node.issues.push({
      severity: 'warn',
      code: 'full-scan',
      message: `${detail} — a full scan, no index narrowed the read`,
    });
  }
  if (detail.includes('USE TEMP B-TREE FOR ')) {
    node.issues.push({ severity: 'warn', code: 'temp-btree', message: detail });
  }
  // D15 (info): a covering index avoids a second lookup into the table itself.
  if (detail.includes('USING COVERING INDEX')) {
    node.issues.push({ severity: 'info', code: 'covering-index', message: detail });
  }
  return node;
}

export function parseSqlitePlan(rows: SqliteExplainRow[]): QueryPlan {
  const nodes = new Map<number, PlanNode>();
  for (const row of rows) nodes.set(row.id, nodeForRow(row.detail));

  const root: PlanNode = { label: 'Query plan', metrics: [], issues: [], children: [] };
  for (const row of rows) {
    const node = nodes.get(row.id);
    if (!node) continue;
    const parent = row.parent !== 0 ? nodes.get(row.parent) : undefined;
    (parent ?? root).children.push(node);
  }

  return {
    kind: 'sqlite',
    root,
    estimatedRowsRead: undefined,
    nativeCost: undefined,
    issues: rollupIssues(root),
    overThreshold: false,
    raw: rows.map((r) => `${r.id}\t${r.parent}\t${r.detail}`).join('\n'),
  };
}

File: apps/kira-studio/internal/queryplan/sqlite.go
package queryplan

import (
	"fmt"
	"regexp"
	"strings"
)

// sqlite.go ports planParsers/sqlite.ts verbatim: SQLite's `EXPLAIN QUERY PLAN` — N rows x 4
// columns (id, parent, notused, detail), no cost and no row estimate whatsoever, so
// EstimatedRowsRead/OverThreshold stay unset/false here (D14). sqlite.org/eqp.html: "The output
// format may change between SQLite releases" — taken as binding (D16): detail is matched on a
// small set of leading tokens for issue detection and otherwise shown verbatim, never re-parsed
// more deeply.

type sqliteExplainRow struct {
	id     int
	parent int
	detail string
}

var sqliteRelationRE = regexp.MustCompile(`^(?:SCAN|SEARCH) (\S+)`)

func sqliteNodeForRow(detail string) Node {
	node := Node{Label: detail, Detail: detail, Metrics: []Metric{}, Issues: []Issue{}, Children: []Node{}}
	if m := sqliteRelationRE.FindStringSubmatch(detail); m != nil {
		node.Relation = m[1]
	}

	// D15: full scan — SCAN (as opposed to SEARCH, which narrows via an index or rowid).
	if strings.HasPrefix(detail, "SCAN ") {
		node.Issues = append(node.Issues, Issue{
			Severity: "warn", Code: "full-scan",
			Message: fmt.Sprintf("%s — a full scan, no index narrowed the read", detail),
		})
	}
	if strings.Contains(detail, "USE TEMP B-TREE FOR ") {
		node.Issues = append(node.Issues, Issue{Severity: "warn", Code: "temp-btree", Message: detail})
	}
	// D15 (info): a covering index avoids a second lookup into the table itself.
	if strings.Contains(detail, "USING COVERING INDEX") {
		node.Issues = append(node.Issues, Issue{Severity: "info", Code: "covering-index", Message: detail})
	}
	return node
}

// sqliteBuilderNode is a pointer-based scratch tree — attachment (a child row arriving, in any
// order, at any depth relative to its parent row) mutates a *sqliteBuilderNode in place. Only
// once every row has been attached does finalizeSqliteNode copy the tree into the value-typed
// Node shape once (plan.go's own note on why a Node value, once copied into a parent's Children
// slice, must never be mutated through a stray pointer afterward — building via pointers first and
// converting exactly once avoids that trap for a tree whose row order does not match its nesting
// order).
type sqliteBuilderNode struct {
	base     Node
	children []*sqliteBuilderNode
}

func finalizeSqliteNode(b *sqliteBuilderNode) Node {
	n := b.base
	n.Children = make([]Node, 0, len(b.children))
	for _, c := range b.children {
		n.Children = append(n.Children, finalizeSqliteNode(c))
	}
	return n
}

func parseSqlitePlan(rows []sqliteExplainRow) Plan {
	builders := make(map[int]*sqliteBuilderNode, len(rows))
	for _, row := range rows {
		builders[row.id] = &sqliteBuilderNode{base: sqliteNodeForRow(row.detail)}
	}

	root := &sqliteBuilderNode{base: Node{Label: "Query plan", Metrics: []Metric{}, Issues: []Issue{}}}
	rawLines := make([]string, 0, len(rows))
	for _, row := range rows {
		rawLines = append(rawLines, fmt.Sprintf("%d\t%d\t%s", row.id, row.parent, row.detail))
		b, ok := builders[row.id]
		if !ok {
			continue
		}
		parent := root
		if row.parent != 0 {
			if p, ok := builders[row.parent]; ok {
				parent = p
			}
		}
		parent.children = append(parent.children, b)
	}

	rootNode := finalizeSqliteNode(root)
	return Plan{
		Kind: "sqlite", Root: rootNode, Issues: rollupIssues(&rootNode), OverThreshold: false,
		Raw: strings.Join(rawLines, "\n"),
	}
}
